core/src/main/scala/neotypes/model.scala

```diff
@@ -0,0 +1,198 @@
+package neotypes
+package model
+
+import mappers.{ParameterMapper, ResultMapper}
+
+import org.neo4j.driver.types.{IsoDuration => NeoDuration, Point => NeoPoint}
+
+import java.time.{LocalDate => JDate, LocalDateTime => JDateTime, LocalTime => JTime, OffsetTime => JZTime, ZonedDateTime => JZDateTime}
+import java.util.{Map => JMap}
+import scala.collection.immutable.ArraySeq
+import scala.jdk.CollectionConverters._
+import scala.util.control.NoStackTrace
+
+/** Safe wrapper over a Neo4j query parameter. */
+final class QueryParam private[neotypes] (private[neotypes] val underlying: AnyRef) extends AnyVal
+object QueryParam {
+  def apply[A](scalaValue: A)(implicit mapper: ParameterMapper[A]): QueryParam =
+    mapper.toQueryParam(scalaValue)
+
+  private[neotypes] def toJavaMap(map: Map[String, QueryParam]): JMap[String, Object] =
+    map.map {
+      case (key, value) =>
+        key -> value.underlying
+    }.asJava
+}
+
+/** Data types supported by Neo4j. */
+object types {
+  /** Parent type of all Neo4j types. */
+  sealed trait NeoType extends Product with Serializable
+
+  /** Parent type of all Neo4j types that have named properties. */
+  sealed trait NeoObject extends NeoType {
+    def properties: Map[String, NeoType]
+
+    final def get(key: String): Option[NeoType] =
+      properties.get(key)
+
+    final def getAs[T](key: String)(implicit mapper: ResultMapper[T]): Either[exceptions.ResultMapperException, T] =
+      properties
+        .get(key)
+        .toRight(left = exceptions.PropertyNotFoundException(key))
+        .flatMap(mapper.decode)
+
+    final def keys: Set[String] =
+      properties.keySet
+
+    final def values: Iterable[NeoType] =
+      properties.values
+  }
+
+  /** Represents a Neo4j heterogeneous list (composite type) */
+  final case class NeoList(values: Iterable[NeoType]) extends NeoType
```

Comment:
Why not a concrete type?

Reply:
Or at least `Seq`, seems like from the description `List` this is supposed to have some actual ordering?

Reply:
Originally I used `List` but that would require unnecessary copying since most of the time this would be created from a **Java** collection using `asScala`.

It was for that reason that I asked the other day if it was an easy way to get an immutable view over some collection without copying it.

I guess I could create my own `Seq` or even `IndexedSeq` out of it.
But, since usually users won't access this directly but `ResultMapper.values`, and that one will call `Map.values` which also returns `Iterable` I decided it would be better to just leave it like that.

Reply:
@armanbilge actually, during the implementation of `Parser` I realized that I needed to produce a concrete collection, so I switched back to `List` 👍